packages/mattermost-redux/src/types/actions.ts

```diff
@@ -12,10 +12,22 @@ export type Thunk = (b: DispatchFunc, a: GetStateFunc) => Promise<ActionResult>
 
 export type Action = GenericAction | Thunk | BatchAction | ActionFunc;
 
-export type ActionResult = {
-    data?: any;
-    error?: any;
+export type ActionResult<Data = any, Error = any> = {
+    data?: Data;
+    error?: Error;
 };
 
 export type DispatchFunc = (action: Action, getState?: GetStateFunc | null) => Promise<ActionResult>;
-export type ActionFunc = (dispatch: DispatchFunc, getState: GetStateFunc) => Promise<ActionResult|ActionResult[]> | ActionResult;
+
+/**
+ * Return type of a redux action.
+ * @usage
+ * ActionFunc<ReturnTypeOfData, ErrorTypeReturnedFromServer>
+ */
+export type ActionFunc<Data = any, Error = any> = (
+    dispatch: DispatchFunc,
+    getState: GetStateFunc
+) => Promise<Data extends any[]
+    ? Array<ActionResult<Data, Error>>
```

Comment:
Not sure if this conditional type is accurate. It says that any time your data returned is an array, you are also returning an array of `ActionResult`s, which doesn't seem like it would always be true. Do we need the conditional type? Can we omit it?

Reply:
That's a good point. That's not true in most cases

Reply:
made the change